src/lib/operators/aggregate_hash.cpp

```diff
@@ -215,54 +215,89 @@ void AggregateHash::_aggregate() {
+                } else {
+                  keys_per_chunk[chunk_id][chunk_offset] = normalize(position.value()) + 1;
+                }
               } else {
-                keys_per_chunk[chunk_id][chunk_offset][group_column_index] = 0u;
+                if (position.is_null()) {
+                  keys_per_chunk[chunk_id][chunk_offset][group_column_index] = 0;
+                } else {
+                  keys_per_chunk[chunk_id][chunk_offset][group_column_index] = normalize(position.value()) + 1;
+                }
               }
-            } else {
-              auto inserted = id_map.try_emplace(position.value(), id_counter);
-              // store either the current id_counter or the existing ID of the value
-              if constexpr (std::is_same_v<AggregateKey, AggregateKeyEntry>) {
-                keys_per_chunk[chunk_id][chunk_offset] = inserted.first->second;
+              ++chunk_offset;
+            });
+          }
+        } else {
+          /*
+          Store unique IDs for equal values in the groupby column (similar to dictionary encoding).
+          The ID 0 is reserved for NULL values. The combined IDs build an AggregateKey for each row.
+          */
+
+          // This time, we have no idea how much space we need, so we take some memory and then rely on the automatic
+          // resizing. The size is quite random, but since single memory allocations do not cost too much, we rather
+          // allocate a bit too much.
+          auto temp_buffer = boost::container::pmr::monotonic_buffer_resource(1'000'000);
+          auto allocator = PolymorphicAllocator<std::pair<const ColumnDataType, AggregateKeyEntry>>{&temp_buffer};
+
+          auto id_map = std::unordered_map<ColumnDataType, AggregateKeyEntry, std::hash<ColumnDataType>,
+                                           std::equal_to<>, decltype(allocator)>(allocator);
+          AggregateKeyEntry id_counter = 1u;
+
+          for (ChunkID chunk_id{0}; chunk_id < chunk_count; ++chunk_id) {
+            const auto chunk_in = input_table->get_chunk(chunk_id);
+            if (!chunk_in) continue;
+
+            const auto base_segment = chunk_in->get_segment(column_id);
+            ChunkOffset chunk_offset{0};
+            segment_iterate<ColumnDataType>(*base_segment, [&](const auto& position) {
+              if (position.is_null()) {
+                if constexpr (std::is_same_v<AggregateKey, AggregateKeyEntry>) {
+                  keys_per_chunk[chunk_id][chunk_offset] = 0u;
+                } else {
+                  keys_per_chunk[chunk_id][chunk_offset][group_column_index] = 0u;
+                }
               } else {
-                keys_per_chunk[chunk_id][chunk_offset][group_column_index] = inserted.first->second;
+                auto inserted = id_map.try_emplace(position.value(), id_counter);
+                // store either the current id_counter or the existing ID of the value
```

Comment:
I don't really understand how this comment is connected to the lines below.

Reply:
Doesn't really help, deleted the comment.